lib/rules/keyframe-selector-notation/index.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+const report = require('../../utils/report');
+const ruleMessages = require('../../utils/ruleMessages');
+const selectorParser = require('postcss-selector-parser');
+const validateOptions = require('../../utils/validateOptions');
+
+const ruleName = 'keyframe-selector-notation';
+
+const messages = ruleMessages(ruleName, {
+	expected: (selector, fixedSelector) => `Expected "${selector}" to be "${fixedSelector}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/list/keyframe-selector-notation',
+};
+
+const PERCENTAGE_SELECTORS = new Set(['0%', '100%']);
+const KEYWORD_SELECTORS = new Set(['from', 'to']);
+
+/** @type {import('stylelint').Rule<'keyword' | 'percentage'>} */
+const rule = (primary, _, context) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: ['keyword', 'percentage'],
+		});
+
+		if (!validOptions) return;
+
+		/**
+		 * @type {Record<primary, {
+		 *   expFunc: (selector: string) => boolean,
+		 *   fixFunc: (selector: string) => string,
+		 * }>}
+		 */
+		const optionFuncs = Object.freeze({
+			keyword: {
+				expFunc: (selector) => KEYWORD_SELECTORS.has(selector),
+				fixFunc: (selector) => {
+					return selector === '0%' ? 'from' : 'to';
+				},
+			},
+			percentage: {
+				expFunc: (selector) => PERCENTAGE_SELECTORS.has(selector),
+				fixFunc: (selector) => {
+					return selector === 'from' ? '0%' : '100%';
+				},
+			},
+		});
+
+		root.walkAtRules(/^(-(moz|webkit)-)?keyframes$/i, (atRuleKeyframes) => {
+			atRuleKeyframes.walkRules((keyframeRule) => {
+				if (keyframeRule.selectors.length > 1) {
+					keyframeRule.selector = selectorParser((selectors) => {
+						selectors.walkTags((selectorTag) => {
+							checkSelector(
+								selectorTag.value,
+								(fixedSelector) => (selectorTag.value = fixedSelector),
+							);
+						});
+					}).processSync(keyframeRule.selector);
+				} else {
+					checkSelector(
+						keyframeRule.selector,
+						(fixedSelector) => (keyframeRule.selector = fixedSelector),
+					);
+				}
```

Comment:
```suggestion
				transformSelector(result, keyframeRule, (selectors) => {
					selectors.walkTags((selectorTag) => {
						checkSelector(
							selectorTag.value,
							(fixedSelector) => (selectorTag.value = fixedSelector),
						);
					});
				});
```

Reply:
Makes sense!

Just to check - I originally added the `if`/`else` logic so that we wouldn't use the selector parser if there was only one selector. I was unsure if the selector parser adds any performance issues on routine checks. Does that matter in the grand scheme of things?

Reply:
I try benchmark, for example:

```console
$ npm run benchmark-rule -- keyframe-selector-notation keyword
```

But I cannot find any performance issues. Can you please try it?

And, I think it useful that `transformSelector()` catches a parser error:

https://github.com/stylelint/stylelint/blob/a6318d45623bf0e4abf80516abc79e8304ea8c01/lib/utils/transformSelector.js#L15